### ISSUE TITLE
getting started: a few hiccups

### ISSUE THREAD

**Issue by faried** [mariusor/littr.go#21]
I didn't want to modify my `pg_hba.conf` rules to allow access to the `postgres` user, so I read the bootstrap code and ran the SQL files manually.  I had to make this change to `db/init.sql`:
```diff
diff --git a/db/init.sql b/db/init.sql
index 8a46b30..784338d 100644
--- a/db/init.sql
+++ b/db/init.sql
@@ -1,7 +1,10 @@
 -- name: drop-tables
+drop table if exists objects;
+drop table if exists votes;
 drop table if exists items;
 drop table if exists accounts;
-drop table if exists votes;
+drop table if exists activities;
+drop table if exists actors;
 drop table if exists instances;
 
 -- name: create-accounts
@@ -102,7 +105,7 @@ create table activities (
 -- this is used to store Note/Article objects that correspond to elements in the items table
 create table objects (
   "id" serial not null constraint objects_pkey primary key,
-  "key" char(32) constraint actors_key_key unique,
+  "key" char(32) constraint objects_key_key unique,
   "pub_id" varchar, -- the activitypub Object ID
   "type" varchar, -- maybe enum
   "url" varchar,
```
and this change to `app/frontend/frontend.go`:
```diff
diff --git a/app/frontend/frontend.go b/app/frontend/frontend.go
index 32d337a..2c44d2e 100644
--- a/app/frontend/frontend.go
+++ b/app/frontend/frontend.go
@@ -188,6 +188,7 @@ func Init(c Config) (handler, error) {
 		account: defaultAccount,
 		sessionKeys: loadEnvSessionKeys(),
 	}
+	c.SessionKeys = h.sessionKeys
 	if c.Logger != nil {
 		h.logger = c.Logger
 	}
```
Using `make run` or `./run.sh main.go` doesn't work: it tries to contact http://localhost/... and of course there's nothing listening there -- the service is running on port 3000.  My `.env` is
```
HOSTNAME=localhost
LISTEN=:3000
ENV=dev
DB_HOST=localhost
DB_NAME=littr
DB_USER=littr
DB_PASSWORD=littr
SESS_AUTH_KEY=FS5PQlDYLeEuGa0F
SESS_ENC_KEY=IQEwbq1wpGNB1FpT
GITHUB_KEY=
GITHUB_SECRET=
FACEBOOK_KEY=
FACEBOOK_SECRET=
GOOGLE_KEY=
GOOGLE_SECRET=
GITLAB_KEY=
GITLAB_SECRET=
#DISABLE_SESSIONS=1
#DISABLE_DOWNVOTING=1
#DISABLE_VOTING=1
REDIS_HOST=localhost
REDIS_PORT=6379
REDIS_PASSWORD=
```
If I modify `app/app.go` to something like
```diff
diff --git a/app/app.go b/app/app.go
index 1069202..878f753 100644
--- a/app/app.go
+++ b/app/app.go
@@ -180,9 +180,17 @@ func loadEnv(l *Application) (bool, error) {
 		l.Secure = false
 	}
 	if l.Secure {
-		l.BaseURL = fmt.Sprintf("https://%s", l.HostName)
+		if l.Port != 443 {
+			l.BaseURL = fmt.Sprintf("https://%s:%d", l.HostName, l.Port)
+		} else {
+			l.BaseURL = fmt.Sprintf("https://%s", l.HostName)
+		}
 	} else {
-		l.BaseURL = fmt.Sprintf("http://%s", l.HostName)
+		if l.Port != 80 {
+			l.BaseURL = fmt.Sprintf("http://%s:%d", l.HostName, l.Port)
+		} else {
+			l.BaseURL = fmt.Sprintf("http://%s", l.HostName)
+		}
 	}
 	if l.Config.Env = EnvType(os.Getenv("ENV")); !validEnv(l.Config.Env) {
 		l.Config.Env = DEV
```
I get the same layout error as in issue #1; I see a loop:
```
time="2019-03-02T20:51:34Z" level=info msg=Started Listen=":3000" env=dev host=localhost
time="2019-03-02T20:51:38Z" level=info msg="[GET http://localhost:3000/self/outbox?context=0&deleted=false&federated=false&maxItems=50&page=1]"
time="2019-03-02T20:51:38Z" level=info msg="[GET http://localhost:3000/self]"
time="2019-03-02T20:51:38Z" level=debug msg="showing self posts"
time="2019-03-02T20:51:38Z" level=info msg="[GET http://localhost:3000/self/outbox?context=0&deleted=false&federated=false&maxItems=50&page=1]"
time="2019-03-02T20:51:38Z" level=info msg="[GET http://localhost:3000/self]"
time="2019-03-02T20:51:38Z" level=debug msg="showing self posts"
time="2019-03-02T20:51:38Z" level=info msg="[GET http://localhost:3000/self/outbox?context=0&deleted=false&federated=false&maxItems=50&page=1]"
time="2019-03-02T20:51:38Z" level=info msg="[GET http://localhost:3000/self]"
time="2019-03-02T20:51:38Z" level=debug msg="showing self posts"
```

**Comment by mariusor** [comment]
Thank you for your feedback @faried. I implemented some the changes where it makes sense. (The ActivityPub tables in the `init.sql` file are not yet used anywhere in the code, I commited them to iterate over the structure a bit) 

One of the problems you encountered is due to the fact that I'm using a load balancer in front of the application (see why below) which listens on port 80 for the host in `HOSTNAME`. I probably should document this. You can change your config:

```sh
HOSTNAME=localhost:3000
```

Besides that I think the problems you encountered running the app could be solved by a recent commit where I started using [godotenv](https://github.com/joho/godotenv). Now if the `.env` or `.env.{dev,qa,prod}` files exist in your run directory, they are loaded without requiring to export all the values beforehand. So now it would suffice to run `make run`

I won't integrate the port changes you made, because I designed this application as a TLS agnostic application, basically relying on the load balancer to handle HTTPS. 

This would allow someone to run multiple applications on different ports in the same environment and round-robin the proxy connections. In the same idea, but on a longer term, the frontend and api packages will evolve into being their own separate applications that will need to be run at the same time. This will most likely be handled by listening on different ports, but I'll see when I get there.

So, in conclusion, please try again with a more recent version (past commit 7d5850f580d4c9e4cb) and let me know what problems you still encounter. 

PS. Even with all these changes there might still be bugs with how some links are generated in the frontend, so user profile links might be broken. :)

**Comment by faried** [comment]
Thanks.  I removed my changes, pulled, and got 0281345c669db45ab42f8dee6026dcc0acba3bc0.  It works now.

Do you have a roadmap for littr?  A list of features you'd like it to have before people can use it publicly?

**Comment by mariusor** [comment]
I have a roadmap in my head, but unfortunately I'm lazy at everything else but programming when it comes to a project.

A pseudo-roadmap can be seen in the [milestones](https://github.com/mariusor/littr.go) section. 

What I would like to get done before doing a "soft launch" is: 

Adding OAuth support to AP endpoint.
Frontend authenticating as OAuth application to AP endpoint.
Full frontend functionality provided through AP (C2S).
Adding federation (S2S) support between littr.me instances for submissions and comments. 
Adding support for some basic moderation mechanisms (reporting, blocking, etc).
